MergerLogic/DataTypes/Fs.cs

```diff
@@ -107,27 +108,28 @@ private IEnumerator<Tile> GetTiles()
             }
         }
 
-        public override List<Tile> GetNextBatch(out string batchIdentifier)
+        public override List<Tile> GetNextBatch(out string batchIdentifier,out string? nextBatchIdentifier, string? inCompletedBatchIdentifier, long? totalTilesCount)
         {
-            batchIdentifier = this._completedTiles.ToString();
-            List<Tile> tiles = new List<Tile>(this.BatchSize);
-
-            if (this._done)
+            lock (_locker)
             {
-                this.Reset();
+                batchIdentifier = this._completedTiles.ToString();
+                List<Tile> tiles = new List<Tile>(this.BatchSize);
+                nextBatchIdentifier = batchIdentifier;
+                if (this._done)
+                {
+                    return tiles;
+                }
+                while (!this._done && tiles.Count < this.BatchSize)
+                {
+                    Tile tile = this._tiles.Current;
+                    tiles.Add(tile);
+                    this._done = !this._tiles.MoveNext();
+                }
+                
+                Interlocked.Add(ref _completedTiles, tiles.Count);
```

Comment:
Why use `Interlocked` if the whole logic is locked?

Reply:
we want to avoid using "+=" when working with threads